### ISSUE TITLE
Make keys editable

### ISSUE THREAD

**Issue by DanielSank** [labrad/scalabrad-web#61]
It does not appear possible to edit registry key values.

**Comment by maffoo** [comment]
Select a key and then press rename. It might be nice to have direct editing like we do for the values, but that can't work with directories (they are links so clicking on them navigates).

**Comment by joshm-3b6b58** [comment]
It works for directories just click where the value would be in the table to select it

**Comment by DanielSank** [comment]
I'm not talking about renaming keys; I want to set the value of a key and this does not work. If I click on a key's row then the row turns grey and that's it.

**Comment by DanielSank** [comment]
When I click on a row for the key `f10` the chrome dev console window shows

```
selectType key selDir null selKey f10
```

but nothing happens.

**Comment by joshm-3b6b58** [comment]
Click on the value of the key and it should be editable. Just start typing. Press "shift+enter" to submit and you should change your key.

It should be clear that the key is editable, the field will have focus and your cursor will apear in the value field

**Comment by DanielSank** [comment]
When I click on the key value the row turns grey and nothing else happens. Typing on the keyboard does nothing. There's no cursor.

**Comment by joshm-3b6b58** [comment]
So in my browser, before clicking anything it looks like this:
![screenshot from 2015-09-24 17 10 38](https://cloud.githubusercontent.com/assets/8173925/10090047/76714c70-62df-11e5-9c47-b3bd2c7af741.png)
When I click on a value (in this case clicking on the string "five2r" it looks like this:
![screenshot from 2015-09-24 17 11 16](https://cloud.githubusercontent.com/assets/8173925/10090053/82312d00-62df-11e5-9965-9f357e6b9c91.png)

Is that not how it works for you?

**Comment by DanielSank** [comment]
@joshmutus Great! Mine does not do that.

**Comment by joshm-3b6b58** [comment]
:/  Are you using Chrome?

**Comment by joshm-3b6b58** [comment]
Did you try refreshing the page? 
Are there any weird compile errors in gulp or sbt?

**Comment by DanielSank** [comment]
@joshmutus 
- I'm using Chrome.
- I checked out the repo this afternoon and am on revision ab30f1709a06deed69fb0642514860a9f7b531e0.
- I am coordinated enough to click on the target area. I have clicked on everything imaginable and nothing gives me the opportunity to edit the item.
- There are no errors in the console window.
- No gulp errors.
- There is an error from the sbt process when I refresh the page:

```
17:14:34.823 [application-akka.actor.default-dispatcher-59] INFO  org.labrad.browser.LabradConnection - disconnected.  will reconnect in 10 seconds
sending: {"jsonrpc":"2.0","method":"org.labrad.disconnected","params":{"host":"redheatdos"}}
java.nio.channels.ClosedChannelException
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.cleanUpWriteBuffer(AbstractNioWorker.java:433)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.writeFromUserCode(AbstractNioWorker.java:128)
    at org.jboss.netty.channel.socket.nio.NioServerSocketPipelineSink.handleAcceptedSocket(NioServerSocketPipelineSink.java:99)
    at org.jboss.netty.channel.socket.nio.NioServerSocketPipelineSink.eventSunk(NioServerSocketPipelineSink.java:36)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:779)
    at org.jboss.netty.channel.Channels.write(Channels.java:725)
    at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.doEncode(OneToOneEncoder.java:71)
    at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.handleDownstream(OneToOneEncoder.java:59)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:591)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:784)
    at org.jboss.netty.channel.SimpleChannelHandler.writeRequested(SimpleChannelHandler.java:292)
    at org.jboss.netty.channel.SimpleChannelHandler.handleDownstream(SimpleChannelHandler.java:254)
    at com.typesafe.netty.http.pipelining.HttpPipeliningHandler.handleDownstream(HttpPipeliningHandler.java:106)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:591)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:582)
    at org.jboss.netty.channel.Channels.write(Channels.java:704)
    at org.jboss.netty.channel.Channels.write(Channels.java:671)
    at org.jboss.netty.channel.AbstractChannel.write(AbstractChannel.java:348)
    at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$play$core$server$netty$PlayDefaultUpstreamHandler$$iteratee$1$1.apply(PlayDefaultUpstreamHandler.scala:305)
    at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$play$core$server$netty$PlayDefaultUpstreamHandler$$iteratee$1$1.apply(PlayDefaultUpstreamHandler.scala:298)
    at play.api.libs.iteratee.Iteratee$$anonfun$flatMap$1$$anonfun$apply$16$$anonfun$apply$17.apply(Iteratee.scala:544)
    at play.api.libs.iteratee.Iteratee$$anonfun$flatMap$1$$anonfun$apply$16$$anonfun$apply$17.apply(Iteratee.scala:544)
    at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
    at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
    at play.api.libs.iteratee.Execution$trampoline$.executeScheduled(Execution.scala:120)
    at play.api.libs.iteratee.Execution$trampoline$.execute(Execution.scala:71)
    at scala.concurrent.impl.Future$.apply(Future.scala:31)
    at scala.concurrent.Future$.apply(Future.scala:492)
    at play.api.libs.iteratee.RunQueue.execute(RunQueue.scala:122)
    at play.api.libs.iteratee.RunQueue.schedule(RunQueue.scala:115)
    at play.api.libs.iteratee.RunQueue.schedule(RunQueue.scala:66)
    at play.api.libs.iteratee.RunQueue.scheduleSimple(RunQueue.scala:88)
    at play.api.libs.iteratee.Concurrent$$anon$6.schedule(Concurrent.scala:179)
    at play.api.libs.iteratee.Concurrent$$anon$6.push(Concurrent.scala:181)
    at play.api.libs.iteratee.Concurrent$Channel$class.push(Concurrent.scala:50)
    at play.api.libs.iteratee.Concurrent$$anon$6.push(Concurrent.scala:176)
    at play.core.actors.WebSocketActor$WebSocketActorSupervisor$$anonfun$receive$1.applyOrElse(WebSocketActor.scala:74)
    at akka.actor.Actor$class.aroundReceive(Actor.scala:467)
    at play.core.actors.WebSocketActor$WebSocketActorSupervisor.aroundReceive(WebSocketActor.scala:25)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
    at akka.actor.ActorCell.invoke(ActorCell.scala:487)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
    at akka.dispatch.Mailbox.run(Mailbox.scala:220)
    at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by joshm-3b6b58** [comment]
Would you mind restarting Chrome?

**Comment by DanielSank** [comment]
@joshmutus restarted Chrome. Nothing.

**Comment by joshm-3b6b58** [comment]
If you restart the scala server does it error out again? To restart, type `ctrl + D` and then `server/run` 

You'll probably have to refresh the browser

**Comment by DanielSank** [comment]
@joshmutus What do you mean "error out"?

**Comment by joshm-3b6b58** [comment]
The scala server _shouldn't_ have that error, if you restart it, does it have that error? 

Even still an error on the scala server shouldn't change how the client is rendering. This is troubling.

**Comment by DanielSank** [comment]
I restarted the scala server. I was able to browse to the registry without seeing errors. However, I still cannot edit key values. Then, if I Ctrl+R the page I see the error reported above.

**Comment by maffoo** [comment]
That error is unrelated; it happens when the connection from the backend scala server to labrad is closed because you refreshed the page. We might be able to clean up the concurrency there, but it's nothing to do with the problem you're having not being able to edit the keys.

**Comment by joshm-3b6b58** [comment]
What's in your `client-js/bower_components`? Can you paste the result of an `ls`?

**Comment by joshm-3b6b58** [comment]
also, what's the output of `$ bower install` ?

**Comment by DanielSank** [comment]
@joshmutus 

```
daniel@princehapi:~/src/scalabrad-web/client-js/bower_components$ ls
accessibility-developer-tools  neon-elements
async                          paper-behaviors
chai                           paper-button
font-roboto                    paper-checkbox
hydrolysis                     paper-dialog
iron-a11y-announcer            paper-dialog-behavior
iron-a11y-keys                 paper-dialog-scrollable
iron-a11y-keys-behavior        paper-drawer-panel
iron-ajax                      paper-elements
iron-autogrow-textarea         paper-fab
iron-behaviors                 paper-header-panel
iron-checked-element-behavior  paper-icon-button
iron-collapse                  paper-input
iron-component-page            paper-item
iron-doc-viewer                paper-material
iron-elements                  paper-menu
iron-fit-behavior              paper-progress
iron-flex-layout               paper-radio-button
iron-form                      paper-radio-group
iron-form-element-behavior     paper-ripple
iron-icon                      paper-scroll-header-panel
iron-icons                     paper-slider
iron-iconset                   paper-spinner
iron-iconset-svg               paper-styles
iron-image                     paper-tabs
iron-input                     paper-toast
iron-jsonp-library             paper-toggle-button
iron-localstorage              paper-toolbar
iron-media-query               paper-tooltip
iron-menu-behavior             platinum-elements
iron-meta                      platinum-push-messaging
iron-overlay-behavior          platinum-sw
iron-pages                     polymer
iron-range-behavior            polymer-ts
iron-resizable-behavior        prism
iron-selector                  prism-element
iron-signals                   promise-polyfill
iron-test-helpers              sinon-chai
iron-validatable-behavior      sinonjs
iron-validator-behavior        stacky
lodash                         sw-toolbox
marked                         test-fixture
marked-element                 web-animations-js
mocha                          webcomponentsjs
neon-animation                 web-component-tester
```

**Comment by DanielSank** [comment]
@joshmutus 

```
daniel@princehapi:~/src/scalabrad-web/client-js$ bower install
bower extra-resolution      Unnecessary resolution: iron-ajax#^1.0.0
bower extra-resolution      Unnecessary resolution: promise-polyfill#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-autogrow-textarea#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-validatable-behavior#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-behaviors#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-a11y-keys-behavior#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-collapse#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-component-page#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-toolbar#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-header-panel#^1.0.0
bower extra-resolution      Unnecessary resolution: marked-element#^1.0.0
bower extra-resolution      Unnecessary resolution: prism-element#^1.0.2
bower extra-resolution      Unnecessary resolution: iron-fit-behavior#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-flex-layout#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-icon#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-icons#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-iconset#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-iconset-svg#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-image#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-input#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-jsonp-library#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-localstorage#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-media-query#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-meta#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-overlay-behavior#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-pages#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-resizable-behavior#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-selector#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-signals#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-test-helpers#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-behaviors#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-button#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-checkbox#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-dialog-scrollable#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-drawer-panel#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-fab#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-icon-button#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-input#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-form-element-behavior#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-item#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-material#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-menu#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-menu-behavior#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-radio-button#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-radio-group#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-ripple#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-spinner#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-tabs#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-toast#^1.0.0
bower extra-resolution      Unnecessary resolution: iron-a11y-announcer#^1.0.0
bower extra-resolution      Unnecessary resolution: paper-toggle-button#^1.0.0
bower extra-resolution      Unnecessary resolution: platinum-sw#~1.0.0
```

**Comment by joshm-3b6b58** [comment]
@DanielSank I am frightened and confused. 

@maffoo Is it worth trying the binary you gave Evan on Dan's machine?

**Comment by joshm-3b6b58** [comment]
Could this be because of the wrong node version?
I am on:

```
$ node -v
v0.12.4
```

**Comment by maffoo** [comment]
I was able to reproduce the issue with a new checkout and fresh npm/bower/jspm install. The problem seems to be that when clicking on the value, the edit field gets shown, but then immediately thereafter a focus-changed event is fired which hides it again. I haven't tracked down which package was updated that caused this change in behavior, but presumably it's again due to the fact that we haven't pinned all the versions, so new checkout are resolving a newer version of some library (see also #63).

I have a fix coming for this specific problem with the edit fields, but it seems like we need to figure out the version pinning for all the packages so that we can make sure everyone who wants to develop of build is seeing a consistent set of packages.

**Comment by maffoo** [comment]
Here are the bower packages that are different between what I had previously, which worked, and what I get after a fresh `bower install`, which does not work:

```
hydrolysis              1.15.4 -> 1.17.0
iron-menu-behavior      1.0.3 -> 1.0.4
iron-overlay-behavior   1.0.6 -> 1.0.7
paper-behaviors         1.0.3 -> 1.0.4
paper-item              1.0.2 -> 1.0.3
paper-progress          1.0.5 -> 1.0.6
paper-tabs              1.0.2 -> 1.0.3
```

**Comment by DanielSank** [comment]
Downgraded typescript and can now edit keys.

**Comment by maffoo** [comment]
Fixed by #66
